Fix the rejected label semantics in webhook metrics, add a counter metrics for webhook rejection with details

roycaihw opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:
The current webhook metrics logs if a webhook `rejected` a request, which could potentially be caused by:
1. an error occurred invoking the webhook (network failure, webhook response is malformed)
2. the webhook is invoked successfully, determines the request should be rejected, returns a rejection with optionally an HTTP status code
3. kube-apiserver hits an internal error

Moreover, the logged "rejection" can be ignored if 1) it's an error invoking the webhook and 2) the webhook is configured to ignore client call failure.

This PR fixes the semantics for the `rejected` label in existing metrics, and adds a new counter metrics `apiserver_admission_webhook_rejection_count` with details about the causing for a rejection.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
The `rejected` label in `apiserver_admission_webhook_admission_duration_seconds` metrices now properly indicates if a request was rejected. Add a new counter metrics `apiserver_admission_webhook_rejection_count` with details about the causing for a webhook rejection.
```

/cc @logicalhan

roycaihw commented:
This is ready for review. PTAL

/cc @logicalhan @liggitt

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/metrics/metrics.go:
```diff
@@ -122,8 +122,8 @@ func newAdmissionMetrics() *AdmissionMetrics {
 	webhook := newMetricSet("webhook",
-		[]string{"name", "type", "operation", "rejected"},
-		"Admission webhook %s, identified by name and broken out for each operation and API resource and type (validate or admit).", false)
+		[]string{"name", "type", "operation", "rejected", "ignore_call_failure", "call_failure", "code"},
```
I don't have a great sense for the extra cost of ~doubling the number of labels. will defer to @logicalhan / @kubernetes/sig-instrumentation-pr-reviews on these

ehashman replied:
Doubling the number of labels isn't necessarily the issue, but how many timeseries they will cause to generate multiplicatively based on their values.

How many values are these labels likely to have? Generating a new timeseries for every HTTP code on top of these other label dimensions, for example, seems like it might result in a lot more load.

roycaihw replied:
- `ignore_call_failure`: boolean, typically only one pre-defined value (true/false) throughout a webhook configuration's life
- `call_failure`: boolean, can be either true (everything works fine) or false (network failure, webhook response is buggy/malformed)
- `code`: int, zero when the request is allowed, non-zero when the request is rejected, depends on webhook implementation-- a well-designed webhook should reject requests with reasonable & standard HTTP status code >= 400 (I imagine 400, 403, 404, 406, 415, 500, 503); a malicious webhook could reject requests with arbitrary numbers
  - perhaps we need to add some validation to bound the value

liggitt replied:
is exact status code correlation important or could we bucket it into 4xx and 5xx?

liggitt replied:
I could envision these response codes:
* 400, 401, 403, 404, 409, 410, 422, 429
* 500, 503

Is ~10 values for the code label problematic? If not, what if we truncated any 4xx code outside that set to 400 and any 5xx code outside that set to 500?

ehashman replied:
@brancz wdyt?

roycaihw replied:
Exact status code correlation is not important to me. My goal is to distinguish a real apiserver internal error from 1) an error invoking the webhook and 2) webhook responding with 500. Bonus point if we can monitor more information with a low cost.

Technically my goal is achievable with one boolean label:
- `(real apiserver)internal_error`: true/false, but having such a label in a webhook metrics sounds weird to me

Other labels that are nice to have:
- `ignore_call_failure` + `call_failure`: let us know when a "rejection" is actually ignored
- `code`: I can go with two values {500, xxx} or three values {4xx, 500, 5xx}; or {4xx, 5xx} if we have the `internal_error` label

logicalhan replied:
I actually like your suggestion of the one additional label. Why not just add `apiserver_internal_error` and call it a day?

logicalhan replied:
> Is ~10 values for the code label problematic? 

Nah, it's fine.

roycaihw replied:
I realized that `apiserver_internal_error` is not sufficient to tell how many 500s in `apiserver_request_count{code="500"}` are caused by webhooks. I changed `code` to log four values: {0, 4xx, 500, 5xx}

liggitt commented:
plumbing looks reasonable, will defer to instrumentation folks on the metrics bits

roycaihw committed:
fix semantics of the rejected label in webhook metrics

when error calling webhook is ignored, do not log the request as
rejected

roycaihw committed:
add webhook rejection metrics

roycaihw committed:
test

roycaihw committed:
wire up the webhook rejection metrics in webhook handlers

roycaihw commented:
fixed the `rejected` label semantics in the existing metrics; added a separate counter metrics with details about why a rejection happened; rebased to adopt the metrics stability framework. @logicalhan @liggitt PTAL

how the metrics looks like after running webhook e2e tests:
```
# HELP apiserver_admission_webhook_rejection_count [ALPHA] Admission webhook rejection count, identified by name and broken out for each admission type (validating or admit) and operation. Additional labels specify an error type (calling_webhook_error or apiserver_internal_error if an error occurred; no_error otherwise) and optionally a non-zero rejection code if the webhook rejects the request with an HTTP status code (honored by the apiserver when the code is greater or equal to 400). Codes greater than 600 are truncated to 600, to keep the metrics cardinality bounded.
# TYPE apiserver_admission_webhook_rejection_count counter
apiserver_admission_webhook_rejection_count{error_type="calling_webhook_error",name="allow-configmap-with-delay-webhook.k8s.io",operation="CREATE",rejection_code="0",type="validating"} 1
apiserver_admission_webhook_rejection_count{error_type="calling_webhook_error",name="deny-unwanted-pod-container-name-and-label.k8s.io",operation="CREATE",rejection_code="0",type="validating"} 1
apiserver_admission_webhook_rejection_count{error_type="calling_webhook_error",name="fail-closed.k8s.io",operation="CREATE",rejection_code="0",type="validating"} 1
apiserver_admission_webhook_rejection_count{error_type="no_error",name="deny-attaching-pod.k8s.io",operation="CONNECT",rejection_code="400",type="validating"} 1
apiserver_admission_webhook_rejection_count{error_type="no_error",name="deny-crd-with-unwanted-label.k8s.io",operation="CREATE",rejection_code="400",type="validating"} 1
apiserver_admission_webhook_rejection_count{error_type="no_error",name="deny-unwanted-configmap-data.k8s.io",operation="CREATE",rejection_code="400",type="validating"} 13
apiserver_admission_webhook_rejection_count{error_type="no_error",name="deny-unwanted-configmap-data.k8s.io",operation="UPDATE",rejection_code="400",type="validating"} 4
apiserver_admission_webhook_rejection_count{error_type="no_error",name="deny-unwanted-custom-resource-data.k8s.io",operation="CREATE",rejection_code="400",type="validating"} 1
apiserver_admission_webhook_rejection_count{error_type="no_error",name="deny-unwanted-custom-resource-data.k8s.io",operation="DELETE",rejection_code="400",type="validating"} 1
apiserver_admission_webhook_rejection_count{error_type="no_error",name="deny-unwanted-pod-container-name-and-label.k8s.io",operation="CREATE",rejection_code="400",type="validating"} 1
apiserver_admission_webhook_rejection_count{error_type="no_error",name="mutating-is-webhook-configuration-ready.k8s.io",operation="CREATE",rejection_code="400",type="admit"} 8
apiserver_admission_webhook_rejection_count{error_type="no_error",name="validating-is-webhook-configuration-ready.k8s.io",operation="CREATE",rejection_code="400",type="validating"} 10
```

logicalhan commented:
Just a few nits on wordage. From the metrics side, I really like this iteration. 

/lgtm

logicalhan commented:
/lgtm

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/webhook/error.go:
```diff
@@ -32,3 +37,32 @@ func (e *ErrCallingWebhook) Error() string {
 }
+
+// ErrWebhookRejection represents a webhook properly rejecting a request.
+type ErrWebhookRejection struct {
```
this reconstructs the status on every call to Error()/RejectionCode()/ToStatusErr(). If we need a wrapper type, can we just have it contain the pre-constructed status?

roycaihw replied:
Error() exists only to satisfy the error interface. I will update the struct to only contain the pre-constructed status

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/webhook/error.go:
```diff
@@ -32,3 +37,32 @@ func (e *ErrCallingWebhook) Error() string {
+}
+
+func (e *ErrWebhookRejection) Error() string {
+	s := webhookerrors.ToStatusErr(e.WebhookName, e.Status)
```
ToStatusErr does not return nil

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -133,7 +133,28 @@ func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr admission.Attrib
+				rejected = true
+				var errorType admissionmetrics.WebhookRejectionErrorType
+				rejectionCode := 0
+				switch err.(type) {
```
this is fairly complex to follow, can we simplify to a top-level switch?

```go
if err != nil {
  switch err := err.(type) {

    case *webhookutil.ErrCallingWebhook:
      if !ignoreClientCallFailures {
        rejected = true
        admissionmetrics.Metrics.ObserveWebhookRejection(hook.Name, "admit", string(versionedAttr.Attributes.GetOperation()), 0, admissionmetrics.WebhookRejectionCallingWebhookError)
      }

    case *webhookutil.ErrWebhookRejection:
      rejected = true
      admissionmetrics.Metrics.ObserveWebhookRejection(hook.Name, "admit", string(versionedAttr.Attributes.GetOperation()), err.Status.ErrStatus.Code, admissionmetrics.WebhookRejectionNoError)

    default:
      rejected = true
      admissionmetrics.Metrics.ObserveWebhookRejection(hook.Name, "admit", string(versionedAttr.Attributes.GetOperation()), 0, admissionmetrics.WebhookRejectionAPIServerInternalError)
  }
}

admissionmetrics.Metrics.ObserveWebhook(time.Since(t), rejected, versionedAttr.Attributes, "admit", hook.Name)
```

liggitt replied:
same comment applies to validation hook

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -164,6 +184,9 @@ func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr admission.Attrib
 			return apierrors.NewInternalError(err)
 		}
+		if rejectionErr, ok := err.(*webhookutil.ErrWebhookRejection); ok {
+			return rejectionErr.ToStatusErr()
```
is there a test ensuring the structured status error from the webhook gets returned from `Admit()`?

roycaihw replied:
yes there is: https://github.com/kubernetes/kubernetes/blob/c80d2c228353e78c9b85bb43bddbf5f2917be4ba/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/plugin_test.go#L170-L171
https://github.com/kubernetes/kubernetes/blob/c80d2c228353e78c9b85bb43bddbf5f2917be4ba/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/plugin_test.go#L145-L147
examples of expected error: webhook rejection with 403, error calling webhook and failclose
https://github.com/kubernetes/kubernetes/blob/c80d2c228353e78c9b85bb43bddbf5f2917be4ba/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/testcase.go#L482-L522

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -249,7 +272,7 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *v1beta
 
 	if !result.Allowed {
-		return false, webhookerrors.ToStatusErr(h.Name, result.Result)
+		return false, &webhookutil.ErrWebhookRejection{WebhookName: h.Name, Status: result.Result}
```
suggest `return false, &webhookutil.ErrWebhookRejection{Status: webhookerrors.ToStatusErr(h.Name, result.Result)}`

liggitt commented:
/lgtm
/approve
/priority important-soon
/milestone v1.16

roycaihw commented:
/area admission-control
